File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        
        try:
            for character in s:
                if character in ['(', '{', '[']:
                    stack.append(character)
                elif character == ')':
                    if stack.pop() != '(':
                        return False
                elif character == '}':
                    if stack.pop() != '{':
                        return False
                elif character == ']':
                    if stack.pop() != '[':
                        return False
        except IndexError:
            return False
                
        if stack:
            return False
        else:
            return True

File: 0020-valid-parentheses/0020-valid-parentheses.kt
class Solution {
    fun isValid(s: String): Boolean {
        if (s.length % 2 == 1) return false

        val stack = Stack<Char>()

        try {
            s.forEach { c ->
                if (c == '(' || c == '{' || c == '[') {
                    stack.push(c)
                } else {
                    stack.pop().let {
                        if (
                            it == '(' && c != ')'
                            || it == '{' && c != '}'
                            || it == '[' && c != ']'
                        ) {
                            throw Exception()
                        }
                    }
                }
            }
        } catch (e: Exception) {
            return false
        }

        return stack.isEmpty()
    }
}